zigHook: init

AndersonTorres opened the pull request:
###### Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [x] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

AndersonTorres commented:
P.S.: Zig eats a lot of memory!

IogaMaster commented:
+1

AndersonTorres commented:
@ofborg build zf

AndersonTorres commented:
@ofborg build zf

AndersonTorres commented:
@ofborg build zf

AndersonTorres commented:
@ofborg build zf

AndersonTorres commented:
Now I am a bit more confident of splitting this!

AndersonTorres commented:
@RaitoBezarius @figsoda ping (because you both gave the initial ideas).

IogaMaster commented:
@AndersonTorres how would use this in #241741?

AndersonTorres commented:
Seriously speaking, these hooks are very independent and self-contained, obviating the `buildZigPackage` framework.

On the other hand, it is too dependent on Bash (or any Bourne shell for that matter). Not a problem, given that Bash is the plumbing of Nixpkgs.

figsoda commented on pkgs/development/compilers/zig/setup-hook.sh:
```diff
@@ -0,0 +1,74 @@
+function zigInstallPhase {
+    runHook preInstall
+
+    if [ -z "${dontAddPrefix-}" ]; then
```
Do you have a use case where this would be helpful?

figsoda commented on pkgs/development/compilers/zig/setup-hook.sh:
```diff
@@ -0,0 +1,74 @@
+
+addEnvHooks "$targetOffset" zigSetGlobalCacheDir
+
+if [ -z "${dontUseZigBuild-}" -a -z "${buildPhase-}" ]; then
```
just a suggestion, I feel like this could be shorter, same suggestion for the other `dontUse` options

```suggestion
if [ -z "${dontZigBuild-}" -a -z "${buildPhase-}" ]; then
```

AndersonTorres replied:
Other similar codes name their variables the same. Indeed my inspirations (meson, ninja, cmake) use vars like `dontUseMesonConfigure`, `dontUseNinjaCheck`...

It's an unwritten standard I want to keep. Also, I like `longButDescriptiveCamelCaseVariable`s.

figsoda commented on pkgs/development/compilers/zig/0.9.1.nix:
the commit message should say `zig_0_9` instead of `zig_0_9_1`

figsoda commented on pkgs/development/compilers/zig/setup-hook.sh:
```diff
@@ -0,0 +1,74 @@
+    buildPhase=zigBuildPhase
+fi
+
+if [ -z "${dontUseZigCheck-}" -a -z "${checkPhase-}" ]; then
```
shellcheck seems to like this better
```suggestion
if [ -z "${dontUseZigCheck-}" ] && [ -z "${checkPhase-}" ]; then
```

AndersonTorres replied:
I shellcheck'd the whole file.

figsoda commented on pkgs/development/compilers/zig/setup-hook.sh:
```diff
@@ -0,0 +1,74 @@
+
+    zigBuildFlags="$zigDefaultBuildFlags $zigBuildFlags"
+
+    echo "zig build flags: $zigBuildFlags ${zigBuildFlagsArray[@]}"
+
+    zig build $zigBuildFlags "${zigBuildFlagsArray[@]}"
```
we should be able to do something like this
```suggestion
    (set -x; zig build $zigBuildFlags "${zigBuildFlagsArray[@]}")
```

AndersonTorres replied:
Good idea!

figsoda replied:
I see that you switched to `echoCmd`, did this approach not work?

AndersonTorres replied:
Well, it worked, however I preferred the traditional approach.

figsoda commented on pkgs/development/compilers/zig/setup-hook.sh:
```diff
@@ -0,0 +1,74 @@
+
+    local zigDefaultBuildFlags="-Drelease-safe -Dcpu=baseline"
+
+    zigBuildFlags="$zigDefaultBuildFlags $zigBuildFlags"
```
this could be an issue if the phase is run multiple times, or if buildPhase is overridden but installPhase is not

AndersonTorres replied:
It will make the shell script a bit bigger.

figsoda reviewed (commented):
as previously suggested, we should include the conversion of some packages (1 or 2 is fine), and documentation

AndersonTorres commented on pkgs/development/compilers/zig/0.10.nix:
```diff
@@ -81,6 +81,10 @@ stdenv.mkDerivation (finalAttrs: {
     runHook postInstallCheck
   '';
 
+  setupHooks = [
```
What would be the difference?

AndersonTorres replied:
You say,  like `wafHook`?

IogaMaster commented:
> Seriously speaking, these hooks are very independent and self-contained, obviating the buildZigPackage framework.
I will close my pr if merged

RaitoBezarius commented:
> > Seriously speaking, these hooks are very independent and self-contained, obviating the buildZigPackage framework.
> > I will close my pr if merged

I'm sorry you went through that effort, I think we should write documentation when people want to build new in-tree builders like those to give them proper direction on the good way to do it.

AndersonTorres committed:
zigHook: init

A setup hook for using the Zig compiler in Nixpkgs.

*Warning*: the setup-hook shell script was linted with shellcheck!

AndersonTorres committed:
doc/hooks/index.md: add zig.section.md

A somewhat short documentation about zigHook.

AndersonTorres commented:
@figsoda any hints about the documentation and its new format?

figsoda commented on doc/hooks/zig.section.md:
```diff
@@ -0,0 +1,35 @@
+[Zig](https://ziglang.org/) is a general-purpose programming language and toolchain for maintaining robust, optimal and reusable software.
+
+In Nixpkgs, `zigHook` overrides the default build, check and install phases.
+
```
would be nice to see an example here

figsoda commented on pkgs/development/compilers/zig/setup-hook.sh:
```diff
@@ -0,0 +1,92 @@
+readonly zigDefaultFlagsArray=("-Drelease-safe=true" "-Dcpu=baseline")
+
+function zigSetGlobalCacheDir {
+
+    ZIG_GLOBAL_CACHE_DIR=$(mktemp -d)
+    export ZIG_GLOBAL_CACHE_DIR
```
```suggestion
    export ZIG_GLOBAL_CACHE_DIR=$(mktemp -d)
```

AndersonTorres replied:
Shellcheck complained.

figsoda commented on pkgs/development/compilers/zig/setup-hook.sh:
```diff
@@ -0,0 +1,92 @@
+
+    echoCmd 'check flags:' "${flagsArray[@]}"
+    zig build test "${flagsArray[@]}"
+    unset flagsArray
```
is this necessary? because `flagsArray` is already `local`

AndersonTorres replied:
I have seen such pattern repeatedly. Maybe it is really unneeded.

AndersonTorres commented on doc/hooks/zig.section.md:
````diff
@@ -0,0 +1,51 @@
+
+## Example code snippet {#example-code-snippet}
+
+```nix
````
@figsoda what do you think?

figsoda commented on doc/hooks/zig.section.md:
````diff
@@ -0,0 +1,51 @@
+## Example code snippet {#example-code-snippet}
+
+```nix
+{ lib
+, stdenv
+/* . . . */
+, zigHook
+}:
+
+stdenv.mkDerivation {
+/* . . . */
+
+   nativeBuildInputs = [
+     zigHook
+   ];
+
+/* . . . */
+}
````
maybe something like this could be better

```suggestion
{ lib
, stdenv
, zigHook
}:

stdenv.mkDerivation {
  # ...

  nativeBuildInputs = [
    zigHook
  ];
  
  zigBuildFlags = [ "-Dman-pages" ];

  dontUseZigCheck = true;

  # ...
}
```

figsoda commented on doc/hooks/zig.section.md:
```diff
@@ -0,0 +1,51 @@
+
+In Nixpkgs, `zigHook` overrides the default build, check and install phases.
+
+## Example code snippet {#example-code-snippet}
+
```
Not sure if we need this

```suggestion
```

AndersonTorres replied:
I prefer to keep it.

figsoda reviewed (approved):
overall lgtm, there are still some design decisions I'm unsure about, but I don't have a strong opinion on any of them

ambroisie commented on pkgs/development/compilers/zig/setup-hook.sh:
```diff
@@ -0,0 +1,90 @@
+
+# This readonly zigDefaultBuildFlagsArray below is meant to avoid CPU feature
+# impurity in Nixpkgs. However, this flagset is "unstable": it is specifically
+# meant to be controlled by the upstream developer, and she is not obliged to
```
Nit and tangential: isn't "they" more inclusive?

Genuinely curious here, I'm not trying to start a flamewar.

RaitoBezarius replied:
'They' is definitely better but not everyone is an native English speaker, so it's understandable.

erikarvstedt commented on pkgs/development/compilers/zig/hook.nix:
```diff
@@ -0,0 +1,16 @@
+  propagatedBuildInputs = [ zig ];
+
+  meta = {
+    description = "A setup hook for using Zig compiler in Nixpkgs";
```
```suggestion
    description = "A setup hook for using the Zig compiler in Nixpkgs";
```

AndersonTorres replied:
I'll keep.

erikarvstedt replied:
Why? It's a grammatical error.
You could also change it to "A setup hook for building Zig programs", which is shorter and more precise.

AndersonTorres replied:
Done.

erikarvstedt commented on pkgs/development/compilers/zig/setup-hook.sh:
```diff
@@ -0,0 +1,90 @@
+readonly zigDefaultFlagsArray=("-Drelease-safe=true" "-Dcpu=baseline")
+
+function zigSetGlobalCacheDir {
+
```
```suggestion
```

erikarvstedt commented on pkgs/development/compilers/zig/setup-hook.sh:
```diff
@@ -0,0 +1,90 @@
+
+# This readonly zigDefaultBuildFlagsArray below is meant to avoid CPU feature
+# impurity in Nixpkgs. However, this flagset is "unstable": it is specifically
+# meant to be controlled by the upstream developer, and she is not obliged to
```
```suggestion
# meant to be controlled by the upstream developers, and they are not obliged to
```

AndersonTorres commented:
Another round and it will be ready to merge.

AndersonTorres commented:
Let's wait arm darwin, then.

AndersonTorres commented:
One full day? I will not wait, sorry.

AndersonTorres committed:
zigHook: remove badPlatforms

Because zig itself doesn't set it by default.
